storage/series/buffer.go

```diff
@@ -77,20 +78,20 @@ func newDatabaseBuffer(drainFn databaseBufferDrainFn, opts Options) databaseBuff
 }
 
 func (b *dbBuffer) Reset() {
-	b.forEachBucketAsc(b.nowFn(), func(bucket *dbBufferBucket, start time.Time) {
+	b.pastMostBucketIdx = b.forEachBucketAscToDrainOrReset(b.nowFn(), func(bucket *dbBufferBucket, start time.Time) {
```

Comment:
Perhaps perform the assignment of `pastMostBucketIdx` inside `forEachBucketAsc` based on a flag?  That way we just need one function not two.

Could even enum-ify this to make it more clean:

```go
type computeReadWriteBucketIdx int

const (
  readBucketIdx computeReadWriteBucketIdx = iota
  computeBucketIdx
  computeAndWriteBucketIdx
)

// ...
now := b.nowFn()
flags := computeAndWriteBucketIdx
b.forEachBucketAsc(now, flags, func(bucket *dbBufferBucket, start time.Time) {

})
```